### DISCUSSION TITLE
Is there a reverse `z.infer` API?

### DISCUSSION THREAD

**Discussion by jensmeindertsma** [colinhacks/zod#667] [Q&A]
I want to validate my \*schemas\* with TypeScript using actual TS data types.

So let's say I have a type `Todo`:
```ts
type Todo = {
  id: number;
};
```
I wrote a validation schema for this type:
```ts
const schema = z.object({
  id: z.number()
})
```
Let's say I make a change to `Todo`:
```ts
type Todo = {
  id: string;
};
```
Now I need to update the validation schema. I would \*love\* to get a TS error now because the TS type and schema are out of sync. I want to achieve this by passing `Todo` to `z.object`. To do that, I need a helper type that transforms my type into a type that fits into a `ZodRawShape`.

So to summarize,  does `zod` expose something that transforms my `{ id: number }` into a `{ id: z.ZodNumber }` type?

**Comment by alii** [comment]
afaik no, but it would be possible to check that the return type of something is equal to the result of `z.infer` which would enable type safety that way. Here's an example: [typescriptlang.org/play](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwF4F84DMoRHA5EiAE1wG4BYAKEoGMIA7AZ3geoAsBTEAQzgF44kAOggAjAFbtqMABQJKASDpcQ7AFwDBTKMDoBzaQEoANJRQHyVCjACeYdnABK7BgFcANvH5Cd6dlAA8NnYQ6HAsHNwAfBaU6C50UsD0cDBcANbOAMpsnFwAYtBOrh7+ACpw7AAeMOx0hAwaAFpEpbbsAIJ01pHS4bnq3nS+AaWRFdW19Y7O7vAA-HDl6nTsAG5+BogKAPTbppSUqRkM2RH5hTMe0gpCohJSsgqKymoadC4gIn6GJhTyZr9zAcrOksjluAUoEVZtc-rdxJIZHI-s8VANNDBtHofgoAZQgRQgA)

Or alternatively, if you are using something like Prisma where you want to check you have schemas for every key on a prisma model, I do it like this

```ts
import {User} from '@prisma/client';
import {z} from 'zod';
import {uuid, date} from './common';

export function ensureKeys<T>() {
	return <X extends Record<keyof T, unknown>>(value: X & Record<never, never>) => value;
}

export function pick<T, K extends Array<keyof T>>(obj: T, keys: K): Pick<T, K[number]> {
	return Object.fromEntries(keys.filter(key => key in obj).map(key => [key, obj[key]])) as Pick<
		T,
		K[number]
	>;
}

export const userSchema = ensureKeys<User>()({
	id: uuid,
	discord_id: z.string(),
	username: z
		.string()
		.max(25)
		.min(2)
		.transform(value => value.toLowerCase()),
	password: z.string().min(6, 'Your password must be at least six characters!'),
	email: z
		.string()
		.email()
		.transform(value => value.toLowerCase()),
	created_at: date,
	updated_at: date,
	website: z.string().url(),
});
```

Now, `userSchema` will have an error if not all properties exist in the passed object. Unfortunately we can't get any stricter than this but that works for the most part.

**Comment by alii** [comment]
update: this is definitely possible, I looked into it a bit more.

```ts
import {z} from 'zod';

type SchemaFor<T> = z.ZodType<T, z.ZodTypeDef, T>;

function schemaFor<T>(schema: SchemaFor<T>) {
	return (data: unknown) => schema.parse(data);
}
```

[Playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwF4F84DMoRHA5EiAE1wG4BYAKEpgE8wBTOAZQGMALekAQwDFoAeACoA+OAF44SAHQAtIoLr0hAGkmz5igCL10qkeSoV0AVwB2LGMAim4AZ3adeAkQAp7HbgC5mD7nyhCwgCUiJQAkFD0MMZQNi6EXDBc3mYA1qYQAO6mIWKi7o5SYFxQtvTxiVxBBiiU1IpwAKplUOKIplwg9N62MFDApgDmKAaUBX4CzfRQwi7h0hAARgBW9BYuCOFhXIPdaqbGIIvTLkHK4ShBlNV1FONOAVMzcxRwb2pLq+ubr+9-HV1vNJev0hqdzr83pdrgYgA)

**Comment by colinhacks** [comment]
There's a long discussion about this concept here: https://github.com/colinhacks/zod/issues/372

The solution I recommend is this, explained in [more detail here](https://github.com/colinhacks/zod/issues/372#issuecomment-826380330).

```ts
const schemaForType = <T>() => <S extends z.ZodType<T, any, any>>(arg: S) => {
  return arg;
};

// use like this:
const dog = schemaForType<Dog>()(
  z.object({
    name: z.string(),
    neutered: z.boolean(),
  })
);
```
